src/Repositories/versionable/VersionableRepository.ts

```diff
@@ -0,0 +1,80 @@
+import * as mongoose from 'mongoose';
+import { DocumentQuery } from 'mongoose';
+import UserRepository from './../../Repositories/user/UserRepository';
+
+export default class VersionableRepository<D extends mongoose.Document, M extends mongoose.Model<D>> {
+    public static generateObjectId() {
+        return String(mongoose.Types.ObjectId());
+    }
+    private modelType: M;
+    constructor(modelType) {
+        this.modelType = modelType;
+    }
+    public async create(options): Promise<D> {   // promise D check,verify and create all schema details
+        const id = await VersionableRepository.generateObjectId();
+        const model = new this.modelType({
+            ...options,
+            _id: id,
+            createdBy: options.userId,
+            originalId: id,
+            updatedBy: options.userId,
+        });
+        return model.save().then((record) => record.toObject());
+    }
+    public async update(id, options) {
+        try {
+            let originalData;
+            const userRepository = new UserRepository();
+            const updateUser = await userRepository.findOne({ originalId: id, deletedAt: { $exists: false } });
+            if (!updateUser) {
+                throw 'not found in update';
+            }
+            else {
+                originalData = updateUser;
+                const id = VersionableRepository.generateObjectId();
+                const modelCreate = new this.modelType({
+                    ...originalData,
+                    ...options,
+                    _id: id,
+                });
+                const record = await this.modelType.create(modelCreate);
+                await record.toObject();
+                const newId = originalData.id;
+                const modelUpdate = new this.modelType({
+                    ...originalData,
+                    deletedAt: Date.now(),
+                });
+                return this.modelType.updateOne({ _id: newId }, modelUpdate);
+            }
+        }
+        catch (error) {
+            throw error;
+        }
+    }
+    public async delete(id) {
+        try {
+            let originalData;
+            const findDelete = await this.modelType.findOne({ originalId: id, deletedAt: { $exists: false } }).lean();
+            if (!findDelete) {
+                throw 'not found in delete';
```

Comment:
Error handling should not be based on a check. Error should be handled inside catch.